RipCurrent/ripcurrent.h

```diff
@@ -0,0 +1,659 @@
+    std::ostringstream os;
+    os << "RipCurrent run context was left hanging. Call `.Sync()`, or, well, something else. -- D.K.\n";
+    T_BASE::GetDefinition().FullDescription(os);
+    return std::move(RipCurrentRunContext(os.str()));
+  }
+
+ private:
+  std::shared_ptr<T_BASE> shared_impl_spawner_;
+};
+
+// Helper code to initialize the next handler in the chain before the user code is constructed.
+// TL;DR: The user should be able to use `emit()` from constructor, no strings attached. Thus,
+// the destination fo this `emit()` should be initialized before the user code is.
+template <InputPolicy INPUT>
+class NextHandlerContainer {
+ public:
+  virtual ~NextHandlerContainer() = default;
+
+  void SetNextHandler(std::shared_ptr<EntriesConsumer<INPUT>> next) const { next_handler_ = next.get(); }
+
+ protected:
+  void emit(int x) const { next_handler_->Accept(x); }
+
+ private:
+  // NOTE(dkorolev): This field is pre-initialized by an instance of another helper class, which is
+  // constructed before the constuctor of the user code, derived indirectly from `NextHandlerContainer`,
+  // is run. This is done to enable user code to do `emit()` from within the constructor, or from callbacks
+  // initialized from within the constructor (for example, HTTP endpoints).
+  // Since such an initialization happens before the constructor has been properly called,
+  // the data field it has to be a POD. An `std::shared_ptr<>` would not work due to being uninitialized first
+  // (and re-initialized over later).
+  // Thus, we use a bare pointer here. Note that the `Instance` classes, that ultimately instantiate user code,
+  // do wrap the top-level `next` into its own `std::shared_ptr<>` and make sure to construct it before (and
+  // destruct it after) the user code itself.
+  mutable EntriesConsumer<INPUT>* next_handler_;
+};
+
+template <InputPolicy INPUT, typename T>
+class NextHandlerInitializer {
+ public:
+  template <typename... ARGS>
+  NextHandlerInitializer(std::shared_ptr<EntriesConsumer<INPUT>> next, ARGS&&... args)
+      : early_initializer_(impl_, next), impl_(std::forward<ARGS>(args)...) {}
+
+  void Accept(const int& x) { impl_.f(x); }  // TODO(dkorolev): This won't be `int` very soon.
+  void Accept(const DummyNoEntryType&) {}
+
+ private:
+  struct NextHandlerEarlyInitializer {
+    NextHandlerEarlyInitializer(NextHandlerContainer<INPUT>& instance_to_initialize,
+                                std::shared_ptr<EntriesConsumer<INPUT>> next) {
+      instance_to_initialize.SetNextHandler(next);
+    }
+  };
+
+  NextHandlerEarlyInitializer early_initializer_;
+  T impl_;
+};
+
+// Base classes for user-defined code, for `is_base_of<>` `static_assert()`-s.
+template <InputPolicy INPUT, OutputPolicy OUTPUT>
+class UserClassTopLevelBase {};
+
+template <InputPolicy INPUT, OutputPolicy OUTPUT, typename USER_CLASS, typename EMITTABLE_TYPES_AS_TYPELIST>
```

Comment:
![emittable vs emitable](https://cloud.githubusercontent.com/assets/2159447/9722243/514027d4-555e-11e5-803a-8e2fa83f3159.png)

Both spellings look acceptable, and I think it's safe to keep this way.